### ISSUE TITLE
Support: Scale from zero behaviour

### ISSUE THREAD

**Issue by Hellisotherpeople** [openfaas/faas#975]
I have a function, in my case, it calculates all prime numbers below N, and returns them to the user. 

**This issue may be human error - but I'm not quite sure how to fix it**

I am doing some testing which involves me curling the endpoint 5 times as quickly as possible, and trying to measure the performance of this function over those 5 invocations. 

## Expected Behaviour
all 5 functions scale up, run, and output their results 

## Current Behaviour
Non-deterministic scaling errors on some functions, specifically: unable to scale function: err: error scaling HTTP code 500. This happens to about 1 function for every 5 (the others succeed), and it seems to have no obvious pattern of which containers will fail 

## Steps to Reproduce (for bugs)

1. Build/Deploy Python function to compute all primes below N 
2. Try to run it with 35000 primes (should take 5-7 seconds) 
3. Curl that endpoint with that request 5 times in a for-loop
4. Output results to a file

## Your Environment

* FaaS-CLI version ( Full output from: `faas-cli version` ): 0.7.5

* Docker version `docker version` (e.g. Docker 17.0.05 ): 18.03.1-ce 

* Are you using Docker Swarm or Kubernetes (FaaS-netes)? Kubernetes, with Helm/Chart

* Operating System and version (e.g. Linux, Windows, MacOS): Fedora 27 

My function has a custom read_timeout and write_timeout set to 60 seconds in the podspec. 

When I deployed OpenFaas, I used all default settings except 
--namespace openfaas \
--set basic_auth=false \
--set functionNamespace=openfaas-fn \
--set faasIdler.dryRun=false \
--set faasIdler.inactivityDuration=1m \
--set faasnetesd.imagePullPolicy=IfNotPresent

**Comment by ivanayov** [comment]
Derek add label: support

**Comment by ivanayov** [comment]
Hi @Hellisotherpeople,

Can you share your function code/configuration so we can reproduce with the same?

Can you please share the logs as well?

Thanks
Ivana

**Comment by alexellis** [comment]
What gateway version are you using? Thanks

**Comment by alexellis** [comment]
It'd be ideal if you could share a git repo we can use to set up your scenario to reproduce any issues you're seeing.

**Comment by martindekov** [comment]
Hello @Hellisotherpeople I was wondering how much memory you have on the node you run OpenFaaS on ? Maybe your node is restarting and that is why you are getting internal server error, correct me if I am wrong but calculating prime numbers takes some memory right? Also could you please provide me with how you wrote your function? Thank you !

**Comment by martindekov** [comment]
Also tested with figlet function invoking it with `hey` 35000 times to see if scaling works as expected and it did that is why I reached the upper conclusion

**Comment by martindekov** [comment]
How I reached those conclusions above

* Test scale zero by deploying figlet from store
```
faas-cli store deploy figlet --label com.openfaas.scale.min=1 com.openfaas.scale.max=5 com.openfaas.scale.zero=true
```
* curl 1 time to trigger idler

* wait idler to scale the function down to 0

* invoke 35 000 with hey the following way:

```
hey -n 35000 <ip_of_my_gateway>/function/figlet
```

* scaled from 0 to 5 without problem and invocations started increasing ...

Deployed my version of prime numbers finder in range and it took 4 minutes eventually breaking (my minikube node does not have enough memory I guess) and it turned in `NotReady` state. 

Tested the same function in browser compiler for python (without handler and request part etc.) and the same function broke the compilers, which made me think the memory on the node might be the problem

**Comment by Hellisotherpeople** [comment]
@martindekov 

Thank you for getting back to me, and apologies for taking so long to follow up on this.

Your advice about needing to have more memory helped - we did not know that it defaults to 2 gigs of ram and 2 CPUs - so changing that  seems to have fixed our problem, but I do not want to close this issue yet as we are still having one other issue. 

When I do a faas test, I only get the number of containers to go between 0 and 1, never up to 5 as expected with cold start. I set maximum replicas to 5, minimum to 0, and the scaling percentage to 100% (so I expect it to instantly scale up to 5 containers), but I only ever see it going between 0 and 1 when I run it as I described above. Is this to be expected? It seems like it is not what is desired here. I figured that 5 curls done in a shell script for loop would be enough to trigger the default "high invocation count" and therefor I'd expect an instant attempt to scale from 0-5, but I never observe this happening.



I have not changed the default "high invocation count" rules in Prometheus, and it doesn't appear that this is necessary but I am trying to figure out if it is or not for what we are doing.

The function I am using: 

`python
def era1(n):
    A = range(2, n + 1)
    B, C = [], A
    while C[0] < math.sqrt(n):  # Condition
        firstElement = C[0]
        # The first number in the list is a prime number. Write this number a list of primes, B.
        B += [firstElement]
        # We use comprehension List to filter multiplies using
        C = [x for x in C if x % firstElement != 0]
    # The numbers in the B list and those left in List A are all primes searched.
    return B+C


print(era1(1000000))

`

**Comment by alexellis** [comment]
The code above is not valid for an OpenFaaS template which may be part of the issue. Please use the official Python template and upload a GitHun repo so we can try to reproduce the reported problem.

Thanks,

Alex

**Comment by Hellisotherpeople** [comment]
I am using the [Python3 Debian template](https://github.com/openfaas-incubator/python3-debian) - I apologize for not mentioning that. I have not modified the template. 

That function works on my machine when copied into a "handler.py" and built using faas-cli.

**Comment by Hellisotherpeople** [comment]
I use a default yaml file except I change the "read_timeout" and "write_timeout" to 60s

**Comment by martindekov** [comment]
Hello @Hellisotherpeople ! 

I am glad I helped with my proposal, one thing that I see is that you have set min replicas to `com.openfaas.scale.min=0` instead of 1, actually you need to set the range 1-5 and set faas-idler to true and dry run to false. I think there was discussion whether or not the additional label should be skipped and just 0-5 in case we want the function to be scaled to 0 but eventually additional label was added.

TL:DR try setting `com.openfaas.scale.min=1` instead of `0` I hope this will resolve your problem.

**Comment by martindekov** [comment]
If the comment above solves your problem, you could open an issue if you have time, (I also can do that just let me know) that the behaviour of the idler is odd in case the min replicas are set to 0.

**Comment by martindekov** [comment]
Hello @Hellisotherpeople I think you want behaviour which can be achieved with service mesh [traffic routing](https://istio.io/docs/reference/config/istio.networking.v1alpha3), which is available feature on top of kubernetes(`Istio` is an example) I am almost certain that the principle which is used by default without service mesh with Kubernetes and Docker swarm  to split the load between containers is round robin which means that every time you have request it will be split to `1 2 3 4 5` `1 2 3 4 5` replica of the container in a loop. 

Note this behaviour is not part of OpenFaaS as the Framework works on top of the orchestrators, the problem you describe is more a problem of Swarm and Kubernetes and as far as I know only Kuberentes has solution for this(don't know about Swarm/haven't heard of such a thing in there), not the Framework itself.

(Hope I understood your question right)

cc @stefanprodan

**Comment by Hellisotherpeople** [comment]
Thank you very much for your assistance with this issue. @martindekov The explanations were very helpful here. I am closing this issue.

**Comment by alexellis** [comment]
I am not convinced that "use Istio" is the answer here.

> I find that some of the functions get executed on containers which were already started up.

This is as designed. The replicas in the cluster will take the load.

Concurrency and load also affect this, a single client calling a function may connect to one replica for efficiency through the routing established in Kubernetes.

The core contributors are evaluating whether it is worth load-balancing at layer 7 (OpenFaaS) vs layer 4 (Kubernetes/Swarm). That reminds me that we should create an issue to track this.

Alex
